### PR TITLE
Adds additional project information to BuildOutputNodes

### PR DESCRIPTION
**Pull request by netonjm**

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs [hunk 1]
```diff
@@ -70,7 +70,11 @@ public override void Process ()
 
 		private void BinLog_BuildStarted (object sender, BuildStartedEventArgs e)
 		{
-			AddNode (BuildOutputNodeType.Build, stringPool.Add (e.Message), stringPool.Add (e.Message), true, e.Timestamp);
+			AddNode (BuildOutputNodeType.Build, 
+			         stringPool.Add (e.Message),
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: Only add the message once to the pool

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs [hunk 2]
```diff
@@ -80,12 +84,26 @@ private void BinLog_BuildFinished (object sender, BuildFinishedEventArgs e)
 
 		private void BinLog_ErrorRaised (object sender, BuildErrorEventArgs e)
 		{
-			AddNode (BuildOutputNodeType.Error, stringPool.Add (e.Message), stringPool.Add (e.Message), false, e.Timestamp);
+			AddNode (BuildOutputNodeType.Error, 
+			         stringPool.Add (e.Message),
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: Same as above.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs [hunk 3]
```diff
@@ -80,12 +84,26 @@ private void BinLog_BuildFinished (object sender, BuildFinishedEventArgs e)
...
+			         e.ProjectFile,
+			         e.LineNumber);
 		}
 
 		private void BinLog_WarningRaised (object sender, BuildWarningEventArgs e)
 		{
-			AddNode (BuildOutputNodeType.Warning, stringPool.Add (e.Message), stringPool.Add (e.Message), false, e.Timestamp);
+			AddNode (BuildOutputNodeType.Warning, 
+			         stringPool.Add (e.Message), 
+			         stringPool.Add (e.Message),
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: Same as above

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs [hunk 4]
```diff
@@ -58,6 +58,10 @@ class BuildOutputNode : TreePosition
 		public virtual bool HasWarnings { get; set; } = false;
 		public virtual bool HasData { get; set; } = false;
 
+		public string File { get; set; }
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs]: Should these be pooled?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs [hunk 5]
```diff
@@ -80,12 +84,26 @@ private void BinLog_BuildFinished (object sender, BuildFinishedEventArgs e)
 
 		private void BinLog_ErrorRaised (object sender, BuildErrorEventArgs e)
 		{
-			AddNode (BuildOutputNodeType.Error, stringPool.Add (e.Message), stringPool.Add (e.Message), false, e.Timestamp);
+			AddNode (BuildOutputNodeType.Error, 
+			         stringPool.Add (e.Message), 
+			         stringPool.Add (e.Message), 
+			         false,
+			         e.Timestamp,
+			         e.File,
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: Should these be pooled?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs [hunk 6]
```diff
@@ -174,7 +192,12 @@ public static void ProcessMessageEvent (this MSBuildOutputProcessor processor, B
 
 			string shortMessage = stringPool.Add (e.Message.Replace ("\n\r", " ").Replace ('\n', ' '));
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: I know this went in earlier, but this is creating 4 strings total. Create a stringbuilder from e.Message, then replace newlines (also, `"\n\r"` should be "\r\n"), then use stringPool.Add (builder).

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs (hunk 5 above)
Comment (rodrmoya) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: yes, I'd say so

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs [hunk 7]
```diff
@@ -80,12 +85,28 @@ private void BinLog_BuildFinished (object sender, BuildFinishedEventArgs e)
...
+			var message = stringPool.Add (e.Message);
+			AddNode (BuildOutputNodeType.Error, 
+			         message, 
+			         message, 
+			         false,
+			         e.Timestamp,
+			         e.File,
+			         e.ProjectFile,
+			         e.LineNumber);
 		}
```
Comment (rodrmoya) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: here, `e.File` and `e.ProjectFile` should be pooled, so replace this with `stringPool.Add (e.File), stringPool.Add (e.ProjectFile)`

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs [hunk 8]
```diff
@@ -80,12 +85,28 @@ private void BinLog_BuildFinished (object sender, BuildFinishedEventArgs e)
...
+			var message = stringPool.Add (e.Message);
+			AddNode (BuildOutputNodeType.Warning, 
+			         message, 
+			         message, 
+			         false,
+			         e.Timestamp,
+			         e.File,
+			         e.ProjectFile,
+			         e.LineNumber);
 		}
```
Comment (rodrmoya) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: same here

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs [hunk 9]
```diff
@@ -115,7 +136,8 @@ private void BinLog_ProjectFinished (object sender, ProjectFinishedEventArgs e)
 
 		private void BinLog_TargetStarted (object sender, TargetStartedEventArgs e)
 		{
-			AddNode (BuildOutputNodeType.Target, stringPool.Add (e.TargetName), stringPool.Add (e.TargetName), true, e.Timestamp);
+			var targetName = stringPool.Add (e.TargetName);
+			AddNode (BuildOutputNodeType.Target, targetName, targetName, true, e.Timestamp);
 		}
```
Comment (rodrmoya) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: this was wrong, it seems, in the branch, the 2nd stringPool.Add should be on `e.Message`, so this should be:

`AddNode (BuildOutputNodeType.Target, stringPool.Add (e.TargetName), stringPool.Add (e.Message), true, e.Timestamp);`

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs [hunk 10]
```diff
@@ -129,8 +151,8 @@ private void BinLog_TaskStarted (object sender, TaskStartedEventArgs e)
 				// <Task Message></Task> are removed, we just display the messages
 				return;
 			}
-
-			AddNode (BuildOutputNodeType.Task, stringPool.Add (e.TaskName), stringPool.Add (e.TaskName), true, e.Timestamp);
+			var taskName = stringPool.Add (e.TaskName);
+			AddNode (BuildOutputNodeType.Task, taskName, taskName, true, e.Timestamp);
 		}
```
Comment (rodrmoya) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: same here, should be `AddNode (BuildOutputNodeType.Task, stringPool.Add (e.TaskName), stringPool.Add (e.Message), true, e.Timestamp);`

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs [hunk 11]
```diff
@@ -174,7 +196,12 @@ public static void ProcessMessageEvent (this MSBuildOutputProcessor processor, B
...
 			string shortMessage = stringPool.Add (e.Message.Replace ("\n\r", " ").Replace ('\n', ' '));
 			processor.AddNode (e.Importance == MessageImportance.Low ? BuildOutputNodeType.Diagnostics : BuildOutputNodeType.Message,
-			                   shortMessage, stringPool.Add (e.Message), false, e.Timestamp);
+			                   shortMessage, stringPool.Add (e.Message),
+			                   false, 
+			                   e.Timestamp,
+			                   e.File,
+			                   e.ProjectFile,
+			                   e.LineNumber);
 		}
```
Comment (rodrmoya) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/MSBuildOutputProcessor.cs]: and same here, `e.File` and `e.ProjectFile` should be pooled

**Pull request merged**